File: crates/klyron_cli/engines/js/engine.js
#!/usr/bin/env node
const { spawnSync } = require("child_process");
const fs = require("fs");
const path = require("path");
const os = require("os");

const MAX_OUTPUT = 1 << 20;

function writeOutput(output) {
  process.stdout.write(JSON.stringify(output) + "\n");
}

function runCmd(cmd, args, input, timeout = 30) {
  const result = spawnSync(cmd, args, {
    input: input,
    timeout: timeout * 1000,
    maxBuffer: MAX_OUTPUT,
    encoding: "utf-8",
    stdio: ["pipe", "pipe", "pipe"],
  });
  const stdout = (result.stdout || "").slice(0, MAX_OUTPUT);
  const stderr = (result.stderr || "").slice(0, MAX_OUTPUT);
  return {
    stdout,
    stderr,
    exit_code: result.status ?? (result.error ? 1 : 0),
    result: result.status === 0 ? "ok" : "failed",
  };
}

function execCode(code, files, filename) {
  if (!code && (!files || files.length === 0)) {
    return { stdout: "", stderr: "No code provided", exit_code: 1, result: "" };
  }
  
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "klyron-js-"));
  try {
    if (files && files.length > 0) {
      for (const f of files) {
        const fp = path.join(tmpDir, f.name);
        fs.mkdirSync(path.dirname(fp), { recursive: true });
        fs.writeFileSync(fp, f.content);
      }
    }
    
    const entry = filename || "main.mjs";
    const fp = path.join(tmpDir, entry);
    fs.mkdirSync(path.dirname(fp), { recursive: true });
    if (code) fs.writeFileSync(fp, code);
    
    const result = runCmd("node", [fp]);
    return result;
  } finally {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }
}

function evalExpr(expr) {
  try {
    const result = eval(expr);
    return { stdout: String(result ?? ""), stderr: "", exit_code: 0, result: JSON.stringify(result) };
  } catch (e) {
    return { stdout: "", stderr: e.message, exit_code: 1, result: "" };
  }
}

function lintCode(code, filename) {
  if (!code) return { stdout: "", stderr: "No code", exit_code: 1, result: "" };
  const tmpFile = path.join(os.tmpdir(), `klyron-js-lint-${Date.now()}.js`);
  try {
    fs.writeFileSync(tmpFile, code);
    // Try eslint first
    const eslintResult = runCmd("npx", ["--yes", "eslint", "--format", "json", tmpFile]);
    if (eslintResult.exit_code === 0 || eslintResult.stdout) {
      return { stdout: "", stderr: eslintResult.stdout || eslintResult.stderr, exit_code: eslintResult.exit_code, result: eslintResult.stdout };
    }
    // Fallback: try node --check
    const checkResult = runCmd("node", ["--check", tmpFile]);
    return { stdout: "", stderr: checkResult.stderr || "Lint passed", exit_code: checkResult.exit_code, result: checkResult.exit_code === 0 ? "No errors" : checkResult.stderr };
  } finally {
    try { fs.unlinkSync(tmpFile); } catch {}
  }
}

function formatCode(code, filename) {
  if (!code) return { stdout: "", stderr: "No code", exit_code: 1, result: "" };
  // Try prettier
  const result = runCmd("npx", ["--yes", "prettier", "--stdin-filepath", filename || "input.js"], code, 15);
  if (result.exit_code === 0 && result.stdout) {
    return { stdout: result.stdout, stderr: "", exit_code: 0, result: "ok" };
  }
  return { stdout: code, stderr: "Prettier not available", exit_code: 0, result: "ok" };
}

function checkProject(projectDir) {
  if (!fs.existsSync(path.join(projectDir, "package.json"))) {
    return { stdout: "", stderr: "No package.json found", exit_code: 1, result: "" };
  }
  const result = runCmd("npx", ["--yes", "tsc", "--noEmit"], null, 60);
  return result;
}

function buildProject(projectDir) {
  if (!fs.existsSync(path.join(projectDir, "package.json"))) {
    return { stdout: "", stderr: "No package.json found", exit_code: 1, result: "" };
  }
  const result = runCmd("npm", ["run", "build"], null, 120);
  return result;
}

function testProject(projectDir) {
  if (!fs.existsSync(path.join(projectDir, "package.json"))) {
    return { stdout: "", stderr: "No package.json found", exit_code: 1, result: "" };
  }
  const result = runCmd("npm", ["test"], null, 120);
  return result;
}

// Main loop
const readline = require("readline");
const rl = readline.createInterface({ input: process.stdin, output: process.stdout, terminal: false });

rl.on("line", (line) => {
  line = line.trim();
  if (!line) return;
  
  let input;
  try { input = JSON.parse(line); } catch (e) { writeOutput({ stdout: "", stderr: "Invalid JSON: " + e.message, exit_code: 1, result: "" }); return; }
  
  const { action, code, args, filename, files, project } = input;
  let output;
  
  switch (action) {
    case "exec":
    case "run":
      output = execCode(code || "", files || [], filename || "");
      break;
    case "file":
      output = execCode("", files || [], filename || args || "");
      break;
    case "eval":
      output = evalExpr(code || "");
      break;
    case "lint":
      output = lintCode(code || "", filename || "input.js");
      break;
    case "format":
      output = formatCode(code || "", filename || "input.js");
      break;
    case "check":
      output = checkProject(project || ".");
      break;
    case "build":
      output = buildProject(project || ".");
      break;
    case "test":
      output = testProject(project || ".");
      break;
    case "ping":
      output = { stdout: "pong", stderr: "", exit_code: 0, result: "ok" };
      break;
    default:
      output = { stdout: "", stderr: "Unknown action: " + action, exit_code: 1, result: "" };
  }
  
  writeOutput(output);
});

File: crates/klyron_cli/engines/py/engine.py
#!/usr/bin/env python3
import sys
import json
import ast
import traceback
import subprocess
import tempfile
import os
import signal
import shutil

MAX_OUTPUT = 1 << 20

def write_output(stdout="", stderr="", exit_code=0, result="", diagnostics=None):
    out = {
        "stdout": stdout,
        "stderr": stderr,
        "exit_code": exit_code,
        "result": result,
    }
    if diagnostics:
        out["diagnostics"] = diagnostics
    sys.stdout.write(json.dumps(out, ensure_ascii=False) + "\n")
    sys.stdout.flush()


def exec_code(code, files=None, filename=""):
    try:
        # Multi-file: write to temp dir and run as subprocess
        if files:
            tmpdir = tempfile.mkdtemp(prefix="klyron_py_")
            try:
                for f in files:
                    fpath = os.path.join(tmpdir, os.path.basename(f["name"]))
                    os.makedirs(os.path.dirname(fpath), exist_ok=True)
                    with open(fpath, "w") as fh:
                        fh.write(f["content"])
                if code:
                    entry = os.path.join(tmpdir, filename or "main.py")
                    with open(entry, "w") as fh:
                        fh.write(code)
                else:
                    entry = os.path.join(tmpdir, files[0]["name"])

                result = subprocess.run(
                    [sys.executable, entry],
                    capture_output=True, text=True, timeout=30
                )
                write_output(
                    stdout=result.stdout[:MAX_OUTPUT],
                    stderr=result.stderr[:MAX_OUTPUT],
                    exit_code=result.returncode,
                    result=result.stdout[:MAX_OUTPUT]
                )
            finally:
                shutil.rmtree(tmpdir, ignore_errors=True)
            return

        # Single code exec with stdout capture
        import io
        old_stdout = sys.stdout
        sys.stdout = captured = io.StringIO()
        try:
            compiled = compile(code, "<exec>", "exec")
            local_ns = {"__builtins__": __builtins__}
            exec(compiled, local_ns)
            stdout = captured.getvalue()
            result_val = local_ns.get("_", "")
            write_output(stdout=stdout, exit_code=0, result=json.dumps(result_val))
        finally:
            sys.stdout = old_stdout
    except SyntaxError as e:
        diag = [{"file": e.filename or "<eval>", "line": e.lineno or 0,
                 "col": e.offset or 0, "message": e.msg, "severity": "error"}]
        write_output(stderr=traceback.format_exc(), exit_code=1, diagnostics=diag)
    except Exception as e:
        write_output(stderr=traceback.format_exc(), exit_code=1)


def exec_file(filename):
    try:
        result = subprocess.run(
            [sys.executable, filename],
            capture_output=True, text=True, timeout=30
        )
        write_output(
            stdout=result.stdout[:MAX_OUTPUT],
            stderr=result.stderr[:MAX_OUTPUT],
            exit_code=result.returncode,
            result=result.stdout[:MAX_OUTPUT]
        )
    except subprocess.TimeoutExpired:
        write_output(stderr="Execution timed out", exit_code=124)
    except Exception as e:
        write_output(stderr=str(e), exit_code=1)


def eval_expr(expr):
    try:
        parsed = ast.parse(expr, mode="eval")
        compiled = compile(parsed, "<eval>", "eval")
        result = eval(compiled)
        write_output(result=json.dumps(result))
    except Exception as e:
        write_output(stderr=str(e), exit_code=1)


def check_code(code):
    try:
        ast.parse(code)
        write_output(result="No syntax errors")
    except SyntaxError as e:
        diag = [{
            "file": e.filename or "<eval>",
            "line": e.lineno or 0,
            "col": e.offset or 0,
            "message": e.msg,
            "severity": "error",
        }]
        write_output(stderr=e.msg, exit_code=1, result="Syntax error",
                     diagnostics=diag)


def pip_install(package, project=None):
    try:
        cmd = [sys.executable, "-m", "pip", "install", package]
        if project:
            cmd += ["--target", os.path.join(project, "vendor")]
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=120)
        write_output(
            stdout=result.stdout[:MAX_OUTPUT],
            stderr=result.stderr[:MAX_OUTPUT],
            exit_code=result.returncode,
            result="ok" if result.returncode == 0 else "failed"
        )
    except Exception as e:
        write_output(stderr=str(e), exit_code=1)


def main():
    signal.signal(signal.SIGPIPE, signal.SIG_IGN)

    while True:
        line = sys.stdin.readline()
        if not line:
            break

        line = line.strip()
        if not line:
            continue

        try:
            input_data = json.loads(line)
        except json.JSONDecodeError as e:
            write_output(stderr=f"Invalid JSON: {e}", exit_code=1)
            continue

        action = input_data.get("action", "")
        code = input_data.get("code", "")
        args = input_data.get("args", "")
        filename = input_data.get("filename", "")
        files = input_data.get("files", [])

        if action in ("exec", "run"):
            exec_code(code, files, filename)
        elif action == "file":
            exec_file(filename or args or code)
        elif action == "eval":
            eval_expr(code)
        elif action in ("check", "typecheck"):
            check_code(code)
        elif action == "pip":
            pip_install(args, input_data.get("project"))
        elif action in ("ping", ""):
            write_output(stdout="pong", result="ok")
        else:
            write_output(stderr=f"Unknown action: {action}", exit_code=1)


if __name__ == "__main__":
    main()
